ktor-network/nix/src/io/ktor/network/selector/SelectUtilsNix.kt

```diff
@@ -102,7 +102,15 @@ internal actual class SelectorHelper {
                 try {
                     selector_pselect(maxDescriptor + 1, readSet, writeSet, errorSet).check()
                 } catch (_: PosixException.BadFileDescriptorException) {
-                    // Thrown if the descriptor was closed.
+                    // Thrown if any of the descriptors was closed.
+                    // This means the sets are undefined so do not rely on their contents.
+                    watchSet.forEach { event ->
+                        if (!isDescriptorValid(event.descriptor)) {
+                            event.fail(IOException("Bad descriptor ${event.descriptor} for ${event.interest}"))
+                            watchSet.remove(event)
+                        }
+                    }
+                    continue
```

Comment:
This is the important part, make sure processSelectedEvents is not called on failure.